### ISSUE TITLE
Multi app instances in one OS scenario

### ISSUE THREAD

**Issue by myonlystarWang** [sysrepo/sysrepo#2258]
Hi Michal,
     My scenario is like this:
     In one machine,run one _netopeer2-server,_so have only one connection with management client.But multiple instances of the same program need to be run, and the configuration data structure is the same(you can think of it as _yang_ model structure), but the configuration data of each instance is different.
     In this case, is there an elegant solution?

BR,
Wang

**Comment by michalvasko** [comment]
Yes, right now you can use only one solution. YANG module should reflect this situation so the whole configuration should probably be a `list` so you can have configuration instance for each managed program instance. In case you cannot change the YANG module, you would still be able to use `schema-mount` with the same effect but since *libyang* does not support it yet, you cannot.

**Comment by myonlystarWang** [comment]
Thanks for rapidly response,michal!
I see the solution what you said.For example,I have a yang module for one program:
```
module app{
    namespace "urn:hysatcom:params:xml:ns:yang:app";
	prefix app;
	description
		"app yang module";
	revision 2020-12-14 {
		description
		    "initial create";
	}
	container config_c{
		config true;
		description
			"The basic system configuration of app";
	        leaf a {
		    type uint16;	
		    description
			   "a";
		}
	        leaf b {
		    type uint16;	
		    description
			   "b";
		}
	} //config_c

	list list_l{
                key "c";
		config true;
		description
			"The list configuration of app";
	        leaf c {
		    type uint16;	
		    description
			   "c";
		}
	        leaf d {
		    type uint16;	
		    description
			   "d";
		}
	} //list_l
}
```
I need to change the container `config_c`to a **list** `config_l` with added key "instanceId" in order to distinguish each managed program instance:
```
	list config_l{
                key "instanceId";
		config true;
		description
			"The basic system configuration of app";
	        leaf instanceId {
		    type uint8;	
		    description
			   "instance id";
		}
	        leaf a {
		    type uint16;	
		    description
			   "a";
		}
	        leaf b {
		    type uint16;	
		    description
			   "b";
		}
	} //config_l
```
But how do I modify the other old `list_l`?Which is the correct one below?
**plan A**:add a leaf `instanceId` to the existed list and get the instance id every time program triggers a data change callback to distinguish each managed program instance.
```
	list list_l{
                key "c";
		config true;
		description
			"The list configuration of app";
	        leaf c {
		    type uint16;	
		    description
			   "c";
		}
	        leaf d {
		    type uint16;	
		    description
			   "d";
		}
	        leaf instanceId {
		    type uint16;	
		    description
			   "instance id";
		}   
	} //list_l
```
**plan B**:A way to make lists nested within lists(I don't know if that's allowed in _libyang_)
```
list list_new{
    key "instanceId";
    config true;
    description
        "The new outer list";

    leaf instanceId{ 
        type uint16;	
        description:
            "instance id";
    }   

	list list_l{
                key "c";
		config true;
		description
			"The inner old list configuration of app";
	        leaf c {
		    type uint16;	
		    description
			   "c";
		}
	        leaf d {
		    type uint16;	
		    description
			   "d";
		}
	} //list_l
}//list_new
```

**Comment by myonlystarWang** [comment]
In this scenario,it brings a place that needs to be changed:
It was originally used `sr_get_change_next` to get every changed leaf,but now we don't know the instance number when get the changed leaf value,so we need to get the entire entry of change data by using `sr_get_change_tree`,right?

**Comment by michalvasko** [comment]
Yeah, I simply meant putting the whole config into another `list`. So solution B, you can have nested lists, naturally.

**Comment by myonlystarWang** [comment]
Ok, thanks for giving directions. 
It's very helpful.

**Comment by myonlystarWang** [comment]
Hi Michal,
I've tried a few things and I don't know if it's the right thing to do.
I created a YANG module as below:
```
module app{
    namespace "urn:hysatcom:params:xml:ns:yang:app";
	prefix app;
	description
		"app yang module";
	revision 2020-12-14 {
		description
		    "initial create";
	}	
	list configList{
            key "instanceId";
            config true;
            description
                "The list configuration of app";
            leaf instanceId{
                type uint16;	
		description
                    "instance id";
            }

            container config1{
		config true;
                 description
			"The basic system configuration of app";
	        leaf a {
		    type uint16;	
		    description
		    	   "a";
	        }
	        leaf b {
		    type uint16;	
		    description
		    	   "b";
	        }
	    } //config1

            container config2{
		config true;
                 description
			"The basic system configuration of app";
	        leaf c {
		    type uint16;	
		    description
		    	   "c";
	        }
	        list d {
		    key "e";	
		    description
		    	   "list d";

	            leaf e {
		        type uint16;	
		        description
		            "e";
	            }

	            leaf f {
		        type uint16;	
		        description
		            "f";
	            }
	        }
	    } //config2
	} 
}
```
What I want to achieve is "Subscribe `config1 `and `config2 `separately and with any value of `instanceId `in `configList`"
So I set the xpath input param of `sr_module_change_subscribe` to `/app:configlist/config1` and `/app:configlist/config2`,it seems ok from the result.But is that the best way?
On the other hand,When I subscribe config2 and the leaf `f` value changed ,I noticed the xpath obtained from callback function has two index expression within.In my application, I need to get both of them,the parsing work has to be done by myself,right?

**Comment by michalvasko** [comment]
Yes, you subscribed correctly with those xpaths. As for the xpath received in the callbacks, you can use functions in `sysrepo/xpath,h` to help you with parsing it. Also, you may simply use `sr_get_change_tree_next()` instead of `sr_get_change_next()` and avoid any parsing at all.

**Comment by myonlystarWang** [comment]
Okey,for example the steps I do:
1. I subscirbe the xpath `/app:configlist/config2`,and change the value of ` leaf f` to 1111 with instanceId = 1 and e = 2 
2. In the callback I can get the output lyd_node from using  `sr_get_change_tree_next() `，`node->schema->name` should be `f`,`node->schema->nodetype` should be `LYS_LEAF`,`(const struct lyd_node_leaf_list *)node->value.uint16` should be 1111.
3. If I want to get the two indexs,`node->parent->child->schema->name` should be `e`,`(const struct lyd_node_leaf_list *)(node->parent->child)->value.uint16` should be 2.
4. How do I get the instanceId?Do I still use the` lyd_node ` above?Like this, should `node->parent->parent->parent->child->schema->name` be `instanceId`?

**Comment by michalvasko** [comment]
Yes, all the points are correct. Every change has all of its parents and if they are lists, with all their keys.

> If the above works, I don't seem to need to use sr_get_data() to get the other sibling leaf of example f,if the list d have them,right?

Based on what I just said, this is not true. For example, you would not get the leaf `c` in this case because it is not needed to uniquely identify your leaf `f`. That is, unless the leaf `c` was also changed, in addition to `c`.

**Comment by myonlystarWang** [comment]
>  For example, you would not get the leaf c in this case because it is not needed to uniquely identify your leaf f. That is, unless the leaf c was also changed, in addition to c.

Emm,I don't quite understand.If the leaf c was not changed during this callback trigger,can I get its value by using `node->parent->parent->child`(Assuming that the leaf `f` has changed this time again )?Or only the values of the changed nodes is provided in `sr_get_change_tree_next()`?

**Comment by michalvasko** [comment]
Like you said, only the values of the changed nodes are provided in `sr_get_change_tree_next()`, along with all their parents (and their keys if they happen to be lists), nothing more.

**Comment by myonlystarWang** [comment]
Okey,I get it.It's very clear.
Thank you very much for your help,michal!

**Comment by myonlystarWang** [comment]
Hi michal,
I ran into some more problems in this scenario.In order for you to reproduce it,I just paste what I'm using.
yang module:
```
module Multi{
    namespace "urn:com:params:xml:ns:yang:Multi";
	prefix Multi;
	
	description
		"yang module";
	
	revision 2020-12-18 {
		description
		    "initial create";
	}

	typedef my-decimal {
         type decimal64 {
             fraction-digits 16;
         }
    }
	 
	container config{
		config true;
		
		description
			"The overall configuration ";
			
		leaf instanceNum{
			type uint8;
			
			description
				"instance id";
		}		

		list InstList{
			key "instanceId";
			
			description
				"all instance configuration";
								
			leaf instanceId {
				type uint8;
				
				description
					"instance id";
			}		
		
			container Cfg {
				config true;
				
				description
					"The algorithm";
					
				leaf l2Time {
					type uint8 {
						range "10..32";
					}
					description
						"The L2 time ";		
				}
				
				leaf smoothFactor {
					type my-decimal ;
					
					description
						"Smoothing factor";			
				}

				list PriWeightList{
					key "Pri";
					
					description
						" priority list";
										
					leaf Pri {
						type uint8;
						
						description
							"priority";
					}		
					
					leaf Weigt {
						type my-decimal;
						
						description
							"weights";
					}	
				}	//PriWeightList			
			} //container Cfg 	
		}//InstList
	}//config	
} //module Multi
```
I successfully installed the module and run application_changes_example with command below:
`./application_changes_example Multi /Multi:config/InstList/Cfg`
**I have to point out there were no data of this module in sysrepo at that time.**
Then I add some data by using:
`sysrepocfg --edit=vim -f json -m lgwrMulti`
and paste:
```
{
  "Multi:config": {
    "InstList": [
      {
        "instanceId": 1,
        "Cfg": {
          "smoothFactor": "0.22",
          "l2Time": 14,
          "PriWeightList": [
            {
              "Pri": 3,
              "PriWeigt": "0.3"
            },
            {
              "Pri": 2,
              "PriWeigt": "0.2"
            },
            {
              "Pri": 1,
              "PriWeigt": "0.5"
            }
          ]
        }
      }
    ]
  }
}
```
But unfortunately，there was no output on the screen. It bothers me a lot.
So I thought about changing the subscription path by using: 
`./application_changes_example Multi /Multi:config/InstList/`
or
`./application_changes_example Multi /Multi:config/`
or simply
`./application_changes_example Multi `
All those worked fine.
So are there any restrictions that I don't know about?The _sysrepo_ version is 1.4.58 in Ubuntu_18.04 VM.
Thanks in advance!

BR,
Wang

**Comment by michalvasko** [comment]
I am not sure why it is not working but please please try using *sysrepo* from `devel`, your version is fairly old now. If you still reproduce it then I can take a look myself.
